src/config.c

```diff
@@ -44,22 +44,22 @@ const struct cfg_section *get_cfg_db() {
 	return cfg_database;
 }
 
-static char *trim(char *s, int force)
+static char *trim(char *s)
 {
 	char *e;
 
-	while (*s && (*s == ' ' || *s == '\t')) s++;
-	if (!*s) return s;
+	while (*s == ' ' || *s == '\t')
+		s++;
+	if (!*s)
+		return s;
 
 	e = s + strlen(s) - 1;
-	while (e >= s && (*e == ' ' || *e == '\t')) e--;
-	/*
-	 * NOTE, if there are trailing spaces, then leave 1 of them. There are
-	 * VALID rules, that need a trailing space like $   i.e. appends space
-	 */
-	if (!force && (*(e+1) == ' ' || *(e+1) =='\t'))
-		++e;
```

Comment:
Was the leaving of a trailing whitespace possibly required to accept some existing Hashcat rulesets? Not an issue in core, which doesn't support those anyway, but could be relevant in jumbo. @jfoug's changes above look like they could invoke undefined behavior - often reading one char beyond end of string - so a correct reimplementation could be needed in jumbo.

Reply:
Oh, right, I was going to revisit that thing!

Reply:
Shouldn't core support a rule like `$ ` or `^ ` though? Not to mention `@ `. Sure, there are ways around it, but why would we trim it?

Reply:
I reverted your changes for now. On a quick look I see no problem with Jim's code, I believe at worst it's reading the trailing null.

Reply:
I'd rather consistently trim trailing whitespace, and never rely on it in any of our rules. "Ways around it" is the way to go, I think. As an alternative, maybe we should stop trimming trailing whitespace at all, rather than leave just one whitespace character in there in special cases? Jim's hack feels like the worst of possible approaches.

Jim's code might be OK in terms of UB - I haven't reviewed it thoroughly enough in full context to know if it is OK or not. My original code from prior to yesterday's commit certainly had UB in it - it'd sometimes move the "e" pointer to one char before the start of string. So reverting my changes isn't the best idea long-term, but I understand we might want to do it for now.

Reply:
Looks like the right-trimming is needed primarily for the use in:

```c
        if (cfg_database && (p = strchr(line, '='))) {
                *p++ = 0;
                cfg_add_param(strlwr(trim(line)), trim(p));
```

where there might be extra whitespace before the `'='` and it shouldn't become part of the parameter name. Maybe we should handle this case specially.

Reply:
I've just committed changes to trim() and its uses in core that should also be suitable for jumbo as-is. Please merge.